README.md
ofxPd
=====
<p align="center">
<img src="https://raw.github.com/danomatika/ofxPd/master/ofxaddons_thumbnail.png"/>
</p>

Copyright (c) [Dan Wilcox](danomatika.com) 2011-2013

BSD Simplified License.

For information on usage and redistribution, and for a DISCLAIMER OF ALL
WARRANTIES, see the file, "LICENSE.txt," in this distribution.

See https://github.com/danomatika/ofxPd for documentation as well as the [OF forum post on ofxPd](http://forum.openframeworks.cc/t/ofxpd/6492)

This project has been supported by the [Frank-Ratchye STUDIO for Creative Inquiry at Carnegie Mellon University](http://studioforcreativeinquiry.org).

Description
-----------

ofxPd is an Open Frameworks addon for running an instance of the Pure Data audio environment within an OpenFrameworks application. Audio, messages, and [MIDI](http://en.wikipedia.org/wiki/Musical_Instrument_Digital_Interface) events can be passed to and from pure data patches and the library is thread safe.

[Pure Data](http://pure-data.info) is a graphical patching environment for audio and multimedia (note: the gui and graphics features are not within the scope of this addon) 

[OpenFrameworks](http://www.openframeworks.cc) is a cross platform open source toolkit for creative coding in C++

Build Requirements
------------------

To use ofxPd, first you need to download and install Open Frameworks. Development is against the latest version of Open Frameworks on github. Checkout a tag if you need an earlier, stable version.

[OF github repository](https://github.com/openframeworks/openFrameworks)

On Mac OSX, you will need to install Xcode from the Mac Developer Tools.

On Linux, you can use the Makefile and/or Codeblocks project files (without the "_win" suffix).

On Win, you will need [Codeblocks+MiniGW](http://www.codeblocks.org/downloads/26) and the [Win Codeblocks OF package](http://www.openframeworks.cc/download). Use the Codeblocks projects files with the "_win" suffix. Also, some versions of CB+MinGW are missing some libraries needed by OF (pthreads, etc). Make sure you've followed the [OF Win Codeblocks setup guide](http://openframeworks.cc/setup/codeblocks).

Installation
------------

Place ofxPd within a folder in the apps folder of the OF dir tree:
<pre>
openframeworks/addons/ofxPd
</pre>

#### Which version to use?

If you are using a stable version (0062, 007, ...) of OpenFrameworks then you want to use a git tag of ofxPd for that version. You can select the tag in the Github "Current Branch" menu or clone and check it out using git.

For example, the following commands will clone ofxPd and switch to the OF 0062 tagged version:
<pre>
git clone git://github.com/danomatika/ofxPd.git
cd ofxPd
git checkout 0062
</pre>

Running the Example Projects
----------------------------

The example projects are in the `pdExample` & `pdExampleIOS` folders.

Project files for the examples are not included so you will need to generate the project files for your operating system and development environment using the OF ProjectGenerator which is included with the OpenFrameworks distribution.

Point the ProjectGenerator to `addons/ofxPd`, change the project name to the **exact** name of the example you want to generate (ie `pdExample`), and make sure to choose `ofxPd` from the addons. Hitting "Generate Project" will populate that example with the project files you will need to build it.

### OSX

Open the Xcode project, select the "pdExample Debug" scheme, and hit "Run".

### Linux

Open the Code::Blocks .cbp and hit F9 to build. Optionally, you can build the example with the Makefile.

To build and run it on the terminal:
<pre>
make
make run
</pre>

### Windows

An example Codeblocks workspace is included. Visual Studio is not currently supported by the libpd source.

PitchShifter
------------

PitchShifter is a simple example application which uses an OF gui to control a pitch shifter within a PD patch. It's in `pitchShifter` and `pitchShifterIOS`. Like the basic example, you will have to generate the project files using the ProjectGenerator. 

How to Create a New ofxPd Project
---------------------------------

_Note: These instructions are for manually creating a new project from an existing ofxPd project and it's project files (aka Xcode, C::B, etc). You do not need to follow these steps if you use the ProjecGenerator app in which case you *may* need to add the C Flags as the PG currently seems to have a problem doing this. See the IDE specific instructions on how to do this._

To develop your own project based on ofxPd, either generate a new project with the ProjectGenerator or generate one of the examples, copy, and rename it. You probably want to put it in your apps folder, for example, after copying:
<pre>
openFrameworks/addons/ofxPd/pdExample/ => openFrameworks/apps/myApps/pdExample/
</pre>

It must be 3 levels down in the openframeworks folder structure.

Then after renaming:
<pre>
openFrameworks/apps/myApps/myPdProject/
</pre>

### For Xcode:

Rename the project in Xcode (do not rename the .xcodeproj file in Finder!): Xcode Menu->Project->Rename

### For Codeblocks (Win & Linux):

Rename the *.cbp and *.workspace files to the same name as the project folder. Open the workspace, readd the renamed project file, and remove the old project.

Adding ofxPd to an Existing Project
-----------------------------------

_Note: These instructions are for manually add ofxPd to an existing project. You do not need to follow these steps if you use the ProjecGenerator app in which case you *may* need to add the C Flags as the PG currently seems to have a problem doing this. See the IDE specific instructions on how to do this._

If you want to add ofxPd to another project, you need to make sure you include the src folder:
<pre>
openFrameworks/addons/ofxPd/src
</pre>

You will also need to include some additional C Flags for building the libpd source:

	-DHAVE_UNISTD_H -DHAVE_ALLOCA_H -DUSEAPI_DUMMY -DPD -DLIBPD_EXTRA

_Note: **-DLIBPD_EXTRA** is optional if you do not need/use the externals in `libpd/pure-data/extra`_

### For Xcode:

* create a new group "ofxPd" 
* drag these directories from ofxpd into this new group: ofxPd/src
* add a search path to: `../../../addons/ofxPd/libs/libpd/pure-data/src` under Targets->YourApp->Build->Header Search Paths (make sure "All" is selected)
* under Targets->YourApp->Build->**Other C Flags** (make sure "All" is selected), add
	<pre>-DHAVE_UNISTD_H -DHAVE_ALLOCA_H -DUSEAPI_DUMMY -DPD -DLIBPD_EXTRA</pre>
  * _Note: Make sure you use Other **C** Flags! Other **C++** Flags will **not** work since libpd is written in C._

### For Linux (Makefiles & Codeblocks):

* edit addons.make in your project folder and add the following line to the end of the file: 
	<pre>ofxPd</pre>

### For Codeblocks (Win):

* add the ofxPd sources to the project:
	* right-click on your project in the project tree
	* select "Add Files Recursively ..."
	* navigate and choose the `ofxPd/src` & `ofx/libs` folder
	* remove `ofxPd/libs/libpd/pure-data/extra/pd~` as it dosen't currently build in Win CB
* add defines, search paths, and libraries to link:
	* right-click on your project in the project tree
	* select "Build options..."
	* make sure the project name is selected in the tree (not release or debug)
	* select the "Compiler settings" tab, add the following to the "#defines" tab:
	<pre>
	HAVE_UNISTD_H
	USEAPI_DUMMY
	LIBPD_EXTRA
	MSW
	PD
	PD_INTERNAL
	WINVER=0x502
	</pre>
	* select the "Search directories" tab, click add the search paths:
	<pre>
	..\\..\\..\addons\ofxPd\src
	..\\..\\..\addons\ofxPd\libs\libpd\pure-data\src
	..\\..\\..\addons\ofxPd\libs\libpd\libpd_wrapper
	</pre>
	* select the Linker settings" tab, add the following libraries:
	<pre>
	pthread
	</pre>
	
Notes
-----

### Audio Interfacing & Debugging Audio Issues

libpd as utilized in ofxPd does not handle any of the audio interfacing itself, but is called via the ofSoundStream system within OpenFrameworks. If you have any issues with the sound interface / audio apis themselves, please log an issue to the [OpenFrameworks Github repo](https://github.com/openframeworks/openFrameworks).

### Sample Rate

The sample rate is set to 44100 when initializing ofxPd in the examples. If your sample rate is higher, the playback pitch will be higher. Make sure the sample rate is the same as your system audio sample rate to hear the correct pitch.

For example: The default sample rate on Mac OSX is 96000. Running the app at 44100 results in double the playback pitch while initing ofxPd at 96000 gives the correct pitch.

### Running App in the Background on iOS

If you're using ofxPd to build an audio app on iOS, you probably want the app to keep running while in the background (aka switching between other apps or going to the home screen). You can enable this in Xcode by clicking on the Project in the project tree, selecting the "Capabilities" tab, and turning on the "Background Modes" switch, then checking "Audio, Airplay and Picture in Picture". Next, Set "Application does not run in background" to NO in the "Info" tab.

### Disabling Automatic Screen Locking on iOS

You may be building an audio app for iOS that you want to run without the automatic screen locking mechanism closing it. You can disable the screen lock timer by adding the following to your ofApp setup() function:

    [[UIApplication sharedApplication] setIdleTimerDisabled:YES];

Bugs & Errors
-------------

### Pitch is off on the iPhone 6S

The iPhone 6S hardware seems to prefer a sample rate of 48000 and calling ofSoundStreamSetup() with 44100 will not change that in versions of OF 0.8.4 and previous. This means ofxPd will be running at 44100 but the audio stream is actually 48000, resulting in a higher pitch coming out of your patches and a lower pitch going in.

The fix is to follow Apple's method of setting the *preferred* sample rate, then grabbing what the *actual* sample rate is afterwards. You can then use this real value in ofSoundStreamSetup() and ofxPd::init(). The pdExampleIOS has been updated to show how to do this. Hopefully, this funtionality will be added to OF in the future.

### File "tr1/memory" not found in Xcode

You just upgraded to OSX 10.9 and Xcode 5 right? The default compiler is now LLVM and you need to rebuild your Xcode project files so OF will build correctly. Ude the ProjectGenerator in the OF 0.8.0 download to regenerate the project:

* choose the _parent folder_ of your project folder
* set the name of the project
* add ofxPd as an add-on
* hit generate

Also note, currently the PG doesn't seem to set the C Flags correctly, so you might have to add them manually. See "Adding ofxPd to an Existing Project" on how to do this.

### Unknown type `t_float`, etc

The compiler doesn't recognize the internal Pd types because it's missing the C Flags needed to build libpd. See the section for your IDE in "Adding ofxPd to an Existing Project" on how to do this. 

### Undefined basic_ostream in XCode

If you get the following [linker error](http://www.openframeworks.cc/forum/viewtopic.php?f=8&t=5344&p=26537&hilit=Undefined+symbol#p26537) in XCode:
<pre>
Undefined symbols: "std::basic_ostream<char, std::char_traits<char> ...
</pre>
you need to change the Base SDK to 10.6: Project > Edit Project Settings

### RtAudio Hang on Exit in 0062

RtAudio will hang on app exit in OF 0062. The only way to fix this is to make a small edit to the OF 0062 core by editing `lib/openFrameworks/sound/ofSoundStream.cpp` and commenting line 143 so close() is not called.

### "verbose" redefinition in Win Codeblocks

Currently, there is a static function in the videoinput lib on Win that conflicts with a #define in the Pure Data sources. The easy fix, until the OF core is updated, is to comment out line 115 in `libs\videoInput\include\videoInput.h`.

Note: This change hasn't been tested while using the ofVideoGrabber yet ... there is a slight chance it may cause a crash, be warned.

### "undefined reference to SetDllDirectory" in Win Codeblocks

Newer versions of libpd check the windows version, so this needs to be set via a define in your codeblocks project. If you have an old ofxPd project, this is not set, hence the error. See "Adding ofxPd to an Existing Project" above, and add the `WINVER=0x502` define to your project.
	
Adding Pure Data external libraries to ofxPd
--------------------------------------------


CHANGELOG
v1.5.3: 2015 Oct

	- updated to libpd 0.8.4
	- ofxPd now reinits itself if the buffer size or number of channels changes in audioIn()/audioOut()

v1.5.2: 2015 Jun

	- updated latest libpd which includes big fix for ringbuffer race condition

v1.5.1: 2015 Apr

	- updated to libpd & pd version 0.46-6 (includes new bob~ external)
	- old polling interface removed since it's no longer part of PdBase, use queued=true in init() & receiveMessages()/receieveMidi()
	- fixed examples not building correctly due to remaining polling interface code
	- fixed build issue due to missing expr~ and d_fft_fftsg.c sources 

v1.5.0: 2015 Jan

	- updated to latest libpd & pd version 0.46-4 (some libpd source files changed, update your projects)
	- included extra externals are now loaded by default (expr, bonk~, fiddle~, etc)
	- updated addons_config.mk to work with upcoming OF 0.9.0 project generator
	- renamed examples and removed example project files (use the project generator)

v1.4.0: 2013 Aug

	- general udpates for OF 0.8.0:
		- moved libpd source into libs/libpd to match ofxAddonsTemplate
		- split example into seperate desktop & ios projects (easier to generate)
		- regenerated project files
	- built-in pd externals now work 
	  (bonk~, choice, fiddle~, loop~, lrshift~, pique~, sigmund~, stdout)
	- added PitchShifter example
	- updated to latest libpd

v1.3.0: 2012 Aug, Sep, Oct

	- updated to latest libpd
	- removed [expr~] due to licensing, should be left up to individual developers
	- added ability to open a new patch instance with an existing patch object
	- changed logging to use "Pd" logging module

v1.2.0: 2012 Jul

	- added support for PdBase event polling in ofxPd

v1.1.0: 2012 Mar

	- updated to latest libpd
	- List::asFloat() & List::asSymbol() are now List::getFloat() & List::getSymbol()

v1.0.0: 2011 Oct, Nov

	- added List building & sending
	- wrapped Pd receiver and types within a "pd" namespace
	- reworked api to better match libpd Java api
		- channel is now first arg in midi functions
		- port is now first arg in raw midi byte functions
		- pgm num range is now 1-128 to match [pgmin]/[pgmout] (was 0-127)
		- destination names are now set when finishing compound messages
		- renamed "sysRT" functions to "sysRt"
		- renamed "getArrayLen" to "getArraySize"
		- renamed "*Source" functions to "subscribe" functions
		- renamed "*subscribe" functions to "receive/ignore" functions
		- renamed "ofxPdListener" class to "PdReceiver"
		- renamed "*Listener" functions to "*Receiver" functions
		- renamed Receiver functions (bangReceived, floatReceived, etc) to match
		  Java Receiver (receiveBang, receiveFloat, etc)
		- split midi receiving into PdMidiReceiver base class
		- renamed "dspOn/dspOff" to "start/stop"
		- added computeAudio function
		- receivers now receive from all subscribed sources by default
		- renamed ofxPdListener and ofxPdTypes source files to PdReceiver and PdTypes
		- renamed "getArraySize" function to "arraySize"
		- renamed "setMaxMsgLength" function to "setMaxMessageLen"
		- renamed "getMaxMsgLength" function to "maxMessgeLen"
		- renamed "getBlockSize" function to "blockSize"
		- renamed "sendMsg" & "finishMsg" functions to "sendMessage" & finishMessage"
		- renamed midi funcs/objects to match java api (ie Note->NoteOn, Ctl->ControlChange, etc)
		- renamed "receivedPrint" function to "print"
	- added receiveMidi/ignoreMidi functions to filter midi channel events for PdMidiReceivers
	- separated base libpd wrapper into PdBase class, now extended by ofxPd
		- added setReceiver and setMidiReceiver functions
		- added audio processing (processRaw, processShort, etc) functions
		- Pd class source files (PdTypes, PdReceiver, etc) have been moved to src/pd/cpp
	- added isInited function
	- now checks if HAVE_UNISTD_H & USEAPI_DUMMY are defined before defining
	- PdBase instance calls now wrapped by a PdContext singleton (global instance until multiple
	  context support is added to libpd)

v0.9.0: 2011 Sep

	- updated to newest libpd
	- added max message length setting
	- merged win codeblocks branch
	- cleaned Xcode projects, removed OF 0062 project
	- tagged OF 0062 version

v0.8.0: 2011 Aug
	
	- added codeblocks project for linux
	- added branch with codeblocks project for windows
	- updated readme with minigw info
	- added pd external libs tutorial to readme

v0.7.0: 2011 Jun-Jul

	- added path handling to addToSearchPath()
	- added search path abs test to example
	- added realtime scope array to example
	- updated OF 007 project
	- updated readme and comments

v0.6.0: 2011 May 29

	- added array access
	- added Patch object
	- cleaned and updated api tests in example
	- simplified some api names to match corresponding pd names (midi)
	- midi channels are now numbered 1-16
	- added keyboard piano to example
	- added delay to example
	- now thread safe on all platforms
	- added BSD license

v0.5.0: 2011 May 26
	
	- added OF 007 iOS project
	- reorganized example into AppCore
	- added ticksPerBuffer to init()
	- updated OF 0062 project
	- audio out working on iOS
	- swapped num in and out channel arg order in init() to match ofSoundStream
	- list and message receiving now working

v0.4.0: 2011 May 22

	- added linux support
	- added mutex for thread safety on linux (required by rtaudio)
	- moved libpd sources to src folder
	- bugfix soundstream hang on exit
	- message event sources and subscriptions are now working

v0.3.0: 2011 May 18

	- fleshed out entire libpd api
	- sound in and out working
	- added stream interface and type structs
	- added receiving control functions
	- updated example
	- bugfix listener bugs and print handling

v0.2.0: 2011 May 13
	
	- added libpd source update script
	- removed 'pd' function name prefix
	- added midinote callback
	- added Xcode OF 0062 project
	- updated Xcode 007 project files
	- removed need for libpd static lib
	- bugfix 0062 hangs on exit

v0.1.0:	Initial version 2011 Jan

	- added Xcode OF 007 beta project file

src/ofxPd.h
/*
 * Copyright (c) 2011 Dan Wilcox <danomatika@gmail.com>
 *
 * BSD Simplified License.
 * For information on usage and redistribution, and for a DISCLAIMER OF ALL
 * WARRANTIES, see the file, "LICENSE.txt," in this distribution.
 *
 * See https://github.com/danomatika/ofxPd for documentation
 *
 * This project uses libpd, copyrighted by Miller Puckette and others using the
 * "Standard Improved BSD License". See the file "LICENSE.txt" in src/pd.
 *
 * See http://gitorious.org/pdlib/pages/Libpd for documentation
 *
 */
#pragma once

#include <map>
#include <set>

#include "PdBase.hpp"

///
/// a Pure Data instance
///
/// references: https://github.com/libpd/libpd/wiki
///
/// note: libpd currently does not support multiple states and it is
///       suggested that you use only one ofxPd object at a time
///
/// also: see PdBase.h in src/pd/cpp for some functions which are not wrapped by
///       ofxPd and PdTypes.h for small Pd C++ Objects
///
/// differences from libpd C api and/or C++ wrapper:
///     - the ofxPd object is thread safe
///     - midi channels are 1-16 to match pd ranges
///     - pgm values are 1-128 to match [pgmin]/[pgmout] in pd
///     - init() takes numOutChannels first to match ofSoundStream
///
class ofxPd : public pd::PdBase, protected pd::PdReceiver, protected pd::PdMidiReceiver {

	public :

		ofxPd();
		virtual ~ofxPd();

	/// \section Initializing Pd

		/// initialize audio resources
		///
		/// set the audio latency by setting the libpd ticks per buffer:
		/// ticks per buffer * lib pd block size (always 64)
		///
		/// ie 4 ticks per buffer * 64 = buffer len of 512
		///
		/// you can call this again after loading patches & setting receivers
		/// in order to update the audio settings
		///
		/// the lower the number of ticks, the faster the audio processing
		/// if you experience audio dropouts (audible clicks), increase the
		/// ticks per buffer
		///
		/// set queued = true to use the built in ringbuffers for message and
		/// midi event passing, you will then need to call receiveMessages() and
		/// receiveMidi() in order to pass messages from the ringbuffers to your
		/// PdReceiver and PdMidiReceiver implementations
		///
		/// the queued ringbuffers are useful when you need to receieve events
		/// on a gui thread and don't want to use locking (aka the mutex)
		///
		bool init(const int numOutChannels, const int numInChannels,
		          const int sampleRate, const int ticksPerBuffer=32,
				  bool queued=false);

		/// clear resources, here for future proofing, currently does nothing
		void clear();

	/// \section Adding Search Paths

		/// add to the pd search path
		/// takes an absolute or relative path (in data folder)
		///
		/// note: fails silently if path not found
		///
		void addToSearchPath(const std::string& path);

		/// clear the current pd search path
		void clearSearchPath();

	/// \section Opening Patches

		/// the pd::Patch class is a wrapper around a unique pointer to an open
		/// instance, giving you access to the $0 value for that instance
		///
		/// if you are going to use multiple instances of a patch, you will need
		/// need to keep a Patch object in order to differentiate between them
		///
		/// see src/pd/cpp/PdTypes.hpp for more info

		/// open a patch file, takes an absolute or relative path (in data folder)
		/// returns a Patch object
		pd::Patch openPatch(const std::string& patch);

		/// open a patch file using the filename and path of an existing patch
		///
		/// // open a patch, don't need pd::Patch instance object
		/// pd.openPatch("apatch.pd", "/some/path");
		///
		/// set the filename within the patch object or use a previously opened
		/// object and create a new instance
		///
		/// // open an instance of "somefile.pd", save the instance in a pd::Patch
		/// Patch p2("somefile.pd", "/some/path");    // set file and path
		/// pd.openPatch(p2);
		///
		/// // open a new instance of "somefile.pd"
		/// Patch p3 = pd.openPatch(p2);
		///
		/// p2 and p3 now refer to 2 different instances of "somefile.pd" and 
		/// p2.dollarZero() & p3.dollarZero() should now return different ids
		///
		pd::Patch openPatch(pd::Patch& patch);

		/// close a patch file, takes the patch's basename (filename without extension),
		/// use this function if you've only opened 1 instance of the given patch
		void closePatch(const std::string& patch);

		/// close a patch file using a Patch object, clears the given Patch object
		/// does not affect other open instances of the same patch
		void closePatch(pd::Patch& patch);

	/// \section Audio Processing Control

		/// start/stop audio processing
		///
		/// note: in general, once started, you won't need to turn off audio processing
		///
		/// shortcuts for [; pd dsp 1( & [; pd dsp 0(
		///
		void computeAudio(bool state);
		void start();
		void stop();

	//// \section Message Receiving

		/// subscribe/unsubscribe to source names from libpd
		///
		/// aka the pd receive name
		///
		/// [r source]
		/// |
		///
		/// note: the global source (aka "") exists by default
		///
		void subscribe(const std::string& source);
		void unsubscribe(const std::string& source);
		bool exists(const std::string& source);
		void unsubscribeAll(); //< receivers will be unsubscribed from *all* sources

		/// process the interal message queue if using the ringbuffer:
		///
		/// internally, libpd will use a ringbuffer to pass messages & midi without
		/// needing to require locking (mutexes) if you call init() with queued = true
		///
		/// call these in your update() loop in order to receive waiting messages
		/// or midi data which are then sent to your PdReceiver & PdMidiReceiver
		///
		/// warning: if you call init() with queued = true and *do not* call these
		///          functions, you will *not* receive any messages!
		///
		/// void receiveMessages(); -> calls PdReceiver
		/// void receiveMidi();     -> calls PdMidiReceiver

		/// add/remove incoming event receiver
		///
		/// receivers automatically receive from *all* subscribed sources
		/// as well as print events
		///
		/// see receive/ignore for specific source receiving control
		///
		void addReceiver(pd::PdReceiver& receiver);
		void removeReceiver(pd::PdReceiver& receiver);
		bool receiverExists(pd::PdReceiver& receiver);
		void clearReceivers();	//< also unsubscribes all receivers

		/// set a receiver to receive/ignore a subscribed source from libpd
		///
		/// receive/ignore using a source name or "" for all sources,
		/// make sure to add the receiver and source first
		///
		/// note: the global source (aka "") is added by default
		/// note: ignoring the global source ignores *all* sources,
		///       so the receiver will not receive any message events,
		///		  but still get print events
		///
		/// also: use negation if you want to plug into all sources but one:
		///
		/// pd.receive(receiver);          // receive from *all*
		/// pd.ignore(receiver, "source"); // ignore "source"
		///
		void receiveSource(pd::PdReceiver& receiver, const std::string& source="");
		void ignoreSource(pd::PdReceiver& receiver, const std::string& source="");
		bool isReceivingSource(pd::PdReceiver& receiver, const std::string& source="");

	/// \section Midi Receiving

		/// add/remove incoming midi event receiver
		///
		/// receivers automatically receive from *all* incoming midi channels
		///
		/// see receive/ignore for specific source receiving control
		///
		void addMidiReceiver(pd::PdMidiReceiver& receiver);
		void removeMidiReceiver(pd::PdMidiReceiver& receiver);
		bool midiReceiverExists(pd::PdMidiReceiver& receiver);
		void clearMidiReceivers();

		/// set a receiver to receive/ignore an incoming midi channel
		///
		/// receive/ignore a specific midi channel or 0 for all channels,
		/// make sure to add the receiver first
		///
		/// note: midi bytes are sent to all receivers
		/// note: the global channel (aka 0) is added by default
		/// note: ignoring the global channel ignores *all* channels,
		///       so the receiver will not receive any midi events except for
		///       midi bytes
		///
		/// also: use negation if you want to plug into all channels but one:
		///
		/// pd.receiveMidi(midiReceiver);   // receive from *all* channels
		/// pd.ignoreMidi(midiReceiver, 2); // ignore channel 2
		///
		void receiveMidiChannel(pd::PdMidiReceiver& receiver, int channel=0);
		void ignoreMidiChannel(pd::PdMidiReceiver& receiver, int channel=0);
		bool isReceivingMidiChannel(pd::PdMidiReceiver& receiver, int channel=0);

	/// \section Sending Functions

		/// messages
		void sendBang(const std::string& dest);
		void sendFloat(const std::string& dest, float value);
		void sendSymbol(const std::string& dest, const std::string& symbol);

		/// compound messages
		///
		/// pd.startMessage();
		/// pd.addSymbol("hello");
		/// pd.addFloat(1.23);
		/// pd.finishList("test");  // "test" is the reciever name in pd
		///
		/// sends [list hello 1.23( -> [r test],
		/// you will need to use the [list trim] object on the reciving end
		///
		/// finishMsg sends a typed message -> [; test msg1 hello 1.23(
		///
		/// pd.startMessage();
		/// pd.addSymbol("hello");
		/// pd.addFloat(1.23);
		/// pd.finishMessage("test", "msg1");
		///
		void startMessage();
		void addFloat(const float value);
		void addSymbol(const std::string& symbol);
		void finishList(const std::string& dest);
		void finishMessage(const std::string& dest, const std::string& msg);

		/// compound messages using the ofxPd List type
		///
		/// List list;
		/// list.addSymbol("hello");
		/// list.addFloat(1.23);
		/// pd.sendList("test", list);
		///
		/// sends [list hello 1.23( -> [r test]
		///
		/// clear the list:
		///
		/// list.clear();
		///
		/// stream operators work as well:
		///
		/// list << "hello" << 1.23;
		/// pd.sendMessage("test", "msg1", list);
		///
		/// sends a typed message -> [; test msg1 hello 1.23(
		///
		void sendList(const std::string& dest, const pd::List& list);
		void sendMessage(const std::string& dest, const std::string& msg, const pd::List& list);

		/// midi
		///
		/// send midi messages, any out of range messages will be silently ignored
		///
		/// number ranges:
		/// channel     1 - 16 * dev# (dev #0: 1-16, dev #1: 17-32, etc)
		/// pitch       0 - 127
		/// velocity    0 - 127
		/// control value   0 - 127
		/// program value   1 - 128
		/// bend value     -8192 - 8191
		/// touch value     0 - 127
		///
		/// note, in pd:
		/// [bendin] takes 0 - 16383 while [bendout] returns -8192 - 8192
		/// [pgmin] and [pgmout] are 1 - 128
		///
		void sendNoteOn(const int channel, const int pitch, const int velocity=64);
		void sendControlChange(const int channel, const int controller, const int value);
		void sendProgramChange(const int channel, const int value);
		void sendPitchBend(const int channel, const int value);
		void sendAftertouch(const int channel, const int value);
		void sendPolyAftertouch(const int channel, const int pitch, const int value);

		/// raw midi bytes
		///
		/// value is a raw midi byte value 0 - 255
		/// port is the raw portmidi port #, similar to a channel
		///
		/// for some reason, [midiin], [sysexin] & [realtimein] add 2 to the port num,
		/// so sending to port 1 in ofxPd returns port 3 in pd
		///
		/// however, [midiout], [sysexout], & [realtimeout] do not add to the port num,
		/// so sending port 1 to [midiout] returns port 1 in ofxPd
		///
		void sendMidiByte(const int port, const int value);
		void sendSysex(const int port, const int value);
		void sendSysRealTime(const int port, const int value);

	/// \section Sending Stream Interface

		/// single messages
		///
		/// pd << Bang("test"); /// "test" is the reciever name in pd
		/// pd << Float("test", 100);
		/// pd << Symbol("test", "a symbol");
		///
		/// compound messages
		///
		/// pd << StartMessage() << 100 << 1.2 << "a symbol" << FinishList("test");
		///
		/// midi
		///
		/// pd << NoteOn(64) << NoteOn(64, 60) << NoteOn(64, 60, 1);
		/// pd << ControlChange(100, 64) << ProgramChange(100, 1) << PitchBend(2000, 1);
		/// pd << Aftertouch(127, 1) << PolyAftertouch(64, 127, 1);
		///
		/// compound raw midi byte stream
		///
		/// pd << StartMidi() << 0xEF << 0x45 << Finish();
		/// pd << StartSysex() << 0xE7 << 0x45 << 0x56 << 0x17 << Finish();
		///
		/// see PdBase.h for function declarations

	/// \section Array Access

		/// get the size of a pd array
		/// returns 0 if array not found
		int arraySize(const std::string& arrayName);

		/// read from a pd array
		///
		/// resizes given vector to readLen, checks readLen and offset
		///
		/// returns true on success, false on failure
		///
		/// calling without setting readLen and offset reads the whole array:
		///
		/// vector<float> array1;
		/// readArray("array1", array1);
		///
		bool readArray(const std::string& arrayName, std::vector<float>& dest,
					   int readLen=-1, int offset=0);

		/// write to a pd array
		///
		/// calling without setting writeLen and offset writes the whole array:
		///
		/// writeArray("array1", array1);
		///
		bool writeArray(const std::string& arrayName, std::vector<float>& source,
						int writeLen=-1, int offset=0);

		/// clear array and set to a specific value
		void clearArray(const std::string& arrayName, int value=0);

	/// \section Utils

		/// has this pd instance been initialized?
		/// bool isInited();
		///
		/// get the blocksize of pd (sample length per channel)
		/// static int blockSize();
		///
		/// get/set the max length of messages and lists, default: 32
		/// void setMaxMsgLength(unsigned int len);
		/// unsigned int maxMsgLength();
		///
		/// see PdBase.h for function declarations
	
		/// get the current ticks per buffer,
		/// updated if the buffer size changes in audioIn or audioOut
		int ticksPerBuffer();
	
		/// get the current sample rate
		int sampleRate();
	
		/// get the number of input channels,
		/// updated if the number changes in audioIn()
		int numInChannels();
	
		/// get the number of output channels,
		/// updated if the number changes in audioOut()
		int numOutChannels();
	
		/// check if ofxPd is currently computing audio
		bool isComputingAudio();

	/// \section Audio Processing Callbacks

		/// audio settings will be inited if the buffersize or number of
		/// channels changes in these callbacks
		///
		/// note: the libpd processing is done in the audioOut callback
		virtual void audioIn(float * input, int bufferSize, int nChannels);
		virtual void audioOut(float * output, int bufferSize, int nChannels);

	protected:

		/// message callbacks
		void print(const std::string& message);
		void receiveBang(const std::string& dest);
		void receiveFloat(const std::string& dest, float value);
		void receiveSymbol(const std::string& dest, const std::string& symbol);
		void receiveList(const std::string& dest, const pd::List& list);
		void receiveMessage(const std::string& dest, const std::string& msg, const pd::List& list);

		/// midi callbacks
		void receiveNoteOn(const int channel, const int pitch, const int velocity);
		void receiveControlChange(const int channel, const int controller, const int value);
		void receiveProgramChange(const int channel, const int value);
		void receivePitchBend(const int channel, const int value);
		void receiveAftertouch(const int channel, const int value);
		void receivePolyAftertouch(const int channel, const int pitch, const int value);
		void receiveMidiByte(const int port, const int byte);

	private:

		int ticks; //< number of ticks per buffer
		int bsize; //< current buffer size aka tbp*blocksize
		int srate; //< current sample rate
		int inChannels, outChannels; //< current num of input & output channels
		bool computing; //< is compute audio on?
	
		float* inputBuffer; //< interleaved input audio buffer

		/// a receiving source's pointer and receivers
		struct Source {

			std::set<pd::PdReceiver*> receivers; //< receivers

			// helper functions
			void addReceiver(pd::PdReceiver* receiver) {
				receivers.insert(receiver);
			}

			void removeReceiver(pd::PdReceiver* receiver) {
				std::set<pd::PdReceiver*>::iterator iter;
				iter = receivers.find(receiver);
				if(iter != receivers.end())
					receivers.erase(iter);
			}

			bool receiverExists(pd::PdReceiver* receiver) {
				if(receivers.find(receiver) != receivers.end())
					return true;
				return false;
			}
		};

		std::set<pd::PdReceiver*> receivers;    //< the receivers
		std::map<std::string,Source> sources;   //< subscribed sources
		                                        //< first object always global

		/// a receiving midi channel's receivers
		struct Channel {

			std::set<pd::PdMidiReceiver*> receivers; //< receivers

			// helper functions
			void addMidiReceiver(pd::PdMidiReceiver* receiver) {
				receivers.insert(receiver);
			}

			void removeMidiReceiver(pd::PdMidiReceiver* receiver) {
				std::set<pd::PdMidiReceiver*>::iterator iter;
				iter = receivers.find(receiver);
				if(iter != receivers.end())
					receivers.erase(iter);
			}

			bool midiReceiverExists(pd::PdMidiReceiver* receiver) {
				if(receivers.find(receiver) != receivers.end())
					return true;
				return false;
			}
		};

		std::set<pd::PdMidiReceiver*> midiReceivers;  //< the midi receivers
		std::map<int,Channel> channels;               //< subscribed channels
		                                              //< first object always global
};
